{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests \n",
    "import pandas as pd\n",
    "from random import sample \n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns                                                             \n",
    "import sys \n",
    "from networkx.algorithms import bipartite\n",
    "from networkx.algorithms.community import greedy_modularity_communities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phage</th>\n",
       "      <th>temperate</th>\n",
       "      <th>cluster</th>\n",
       "      <th>subcluster</th>\n",
       "      <th>morphotype</th>\n",
       "      <th>host genus</th>\n",
       "      <th>host species</th>\n",
       "      <th>genome length</th>\n",
       "      <th>is annotated</th>\n",
       "      <th>is phamerated</th>\n",
       "      <th>gcpercent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20ES</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>A2</td>\n",
       "      <td>SIPHO</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>smegmatis</td>\n",
       "      <td>53124</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>63.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>244</td>\n",
       "      <td>True</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIPHO</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>smegmatis</td>\n",
       "      <td>74483</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>63.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32HC</td>\n",
       "      <td>True</td>\n",
       "      <td>Z</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIPHO</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>smegmatis</td>\n",
       "      <td>50781</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>65.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>39HC</td>\n",
       "      <td>False</td>\n",
       "      <td>B</td>\n",
       "      <td>B6</td>\n",
       "      <td>SIPHO</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>smegmatis</td>\n",
       "      <td>71565</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40AC</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>A17</td>\n",
       "      <td>SIPHO</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>smegmatis</td>\n",
       "      <td>53396</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>63.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  phage temperate cluster subcluster morphotype     host genus host species  \\\n",
       "0  20ES      True       A         A2      SIPHO  Mycobacterium    smegmatis   \n",
       "1   244      True       E        NaN      SIPHO  Mycobacterium    smegmatis   \n",
       "2  32HC      True       Z        NaN      SIPHO  Mycobacterium    smegmatis   \n",
       "3  39HC     False       B         B6      SIPHO  Mycobacterium    smegmatis   \n",
       "4  40AC      True       A        A17      SIPHO  Mycobacterium    smegmatis   \n",
       "\n",
       "   genome length  is annotated  is phamerated  gcpercent  \n",
       "0          53124         False           True       63.4  \n",
       "1          74483          True           True       63.4  \n",
       "2          50781         False           True       65.7  \n",
       "3          71565         False           True       70.0  \n",
       "4          53396         False           True       63.3  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_phages = pd.read_csv(\"data/phage_metadata.csv\")\n",
    "df_phages.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene ID</th>\n",
       "      <th>pham</th>\n",
       "      <th>function</th>\n",
       "      <th>translation</th>\n",
       "      <th>orientation</th>\n",
       "      <th>phage</th>\n",
       "      <th>gene number</th>\n",
       "      <th>uncleaned function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20ES_CDS_1</td>\n",
       "      <td>36676</td>\n",
       "      <td>NKF</td>\n",
       "      <td>MYGTRSSAFWASQPGKFDVLNLRMTFPSTSAHEIPDLTATDFVPEN...</td>\n",
       "      <td>F</td>\n",
       "      <td>20ES</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20ES_CDS_10</td>\n",
       "      <td>39578</td>\n",
       "      <td>lysin b</td>\n",
       "      <td>MSLQVGSSGELVNRWIRVMKARFASYAGKLKEDGYFGLDDKAVQQE...</td>\n",
       "      <td>F</td>\n",
       "      <td>20ES</td>\n",
       "      <td>10</td>\n",
       "      <td>lysin b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20ES_CDS_11</td>\n",
       "      <td>34196</td>\n",
       "      <td>terminase</td>\n",
       "      <td>MSLENHHPELAPSPPHIIGPSWQRTVDGSWHLPDPKMTLGWGVLKW...</td>\n",
       "      <td>F</td>\n",
       "      <td>20ES</td>\n",
       "      <td>11</td>\n",
       "      <td>terminase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20ES_CDS_12</td>\n",
       "      <td>39511</td>\n",
       "      <td>portal protein</td>\n",
       "      <td>MTAPLPGQEEIPDPAIARDEMISAFDDAVKNLKINTSYYEAERRPE...</td>\n",
       "      <td>F</td>\n",
       "      <td>20ES</td>\n",
       "      <td>12</td>\n",
       "      <td>portal protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20ES_CDS_13</td>\n",
       "      <td>21454</td>\n",
       "      <td>capsid maturation protease</td>\n",
       "      <td>MITAAVAAYVQRFASMFTGPALSLGEWARFLQTLFPEVQRRYAQAA...</td>\n",
       "      <td>F</td>\n",
       "      <td>20ES</td>\n",
       "      <td>13</td>\n",
       "      <td>capsid maturation protease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene ID   pham                    function  \\\n",
       "0   20ES_CDS_1  36676                         NKF   \n",
       "1  20ES_CDS_10  39578                     lysin b   \n",
       "2  20ES_CDS_11  34196                   terminase   \n",
       "3  20ES_CDS_12  39511              portal protein   \n",
       "4  20ES_CDS_13  21454  capsid maturation protease   \n",
       "\n",
       "                                         translation orientation phage  \\\n",
       "0  MYGTRSSAFWASQPGKFDVLNLRMTFPSTSAHEIPDLTATDFVPEN...           F  20ES   \n",
       "1  MSLQVGSSGELVNRWIRVMKARFASYAGKLKEDGYFGLDDKAVQQE...           F  20ES   \n",
       "2  MSLENHHPELAPSPPHIIGPSWQRTVDGSWHLPDPKMTLGWGVLKW...           F  20ES   \n",
       "3  MTAPLPGQEEIPDPAIARDEMISAFDDAVKNLKINTSYYEAERRPE...           F  20ES   \n",
       "4  MITAAVAAYVQRFASMFTGPALSLGEWARFLQTLFPEVQRRYAQAA...           F  20ES   \n",
       "\n",
       "   gene number          uncleaned function  \n",
       "0            1                         NaN  \n",
       "1           10                     lysin b  \n",
       "2           11                   terminase  \n",
       "3           12              portal protein  \n",
       "4           13  capsid maturation protease  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes = pd.read_csv(\"data/cleaned_gene_list.csv\")\n",
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# phages = sorted(list(df_phages[df_phages[\"cluster\"]==\"A\"][\"phage\"].unique()))\n",
    "phages = df_phages[\"phage\"].unique()\n",
    "gene_identifier = \"function\"\n",
    "\n",
    "labels = [\"5' start\"] + [ident for ident in np.sort(df_genes[gene_identifier].unique())] + [\"3' end\"]\n",
    "label_to_index = { labels[i]:i for i in range(len(labels))}\n",
    "\n",
    "transition_counts = np.zeros((len(labels),len(labels)))\n",
    "\n",
    "for phage in phages:\n",
    "    phage_df = df_genes[df_genes[\"phage\"] == phage].sort_values(by=['gene number'])\n",
    "    i = label_to_index[\"5' start\"]\n",
    "    for index, row in phage_df.iterrows(): \n",
    "        j = label_to_index[row[gene_identifier]]\n",
    "        transition_counts[i][j] += 1\n",
    "        i = j\n",
    "    transition_counts[i][label_to_index[\"3' end\"]] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "transition_matrix = np.zeros((len(labels),len(labels)))\n",
    "for i in range(len(labels)):\n",
    "    sum_row = np.sum(transition_counts[i][:])\n",
    "    transition_matrix[i][:] = transition_counts[i][:]/(sum_row if sum_row > 0 else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(labels)):\n",
    "    for j in range(len(labels)):\n",
    "        if transition_matrix[i][j] == 1:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "G_markov = nx.from_numpy_matrix(transition_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GROUP  0\n",
      "serine homologous recombinase\n",
      "NKF\n",
      "adp-ribosyl glycohydrolase\n",
      "purple acid phosphatase\n",
      "arsenate reductase\n",
      "dpda-like trna-guanine transglycosylase\n",
      "dsdna break-binding protein, adda-like\n",
      "lipase, lipc-like\n",
      "pentapeptide repeat protein\n",
      "pe/ppe family protein\n",
      "trna-methyltransferase\n",
      "tyrosine homologous recombinase\n",
      "\n",
      "\n",
      "GROUP  1\n",
      "holin\n",
      "lysin a, glycosyl hydrolase domain\n",
      "lysin a, l-ala-d-glu peptidase domain\n",
      "lysin a, n-acetylmuramoyl-l-alanine amidase domain\n",
      "lysin a, protease c39 domain\n",
      "lysin a, protease domain\n",
      "lysin a, protease m15 domain\n",
      "lysin a, protease m23 domain\n",
      "\n",
      "\n",
      "GROUP  2\n",
      "major tail protein\n",
      "tail assembly chaperone\n",
      "tail fiber\n",
      "capsid decoration protein\n",
      "glycoside hydrolase\n",
      "head fiber protein\n",
      "\n",
      "\n",
      "GROUP  3\n",
      "adenylate kinase\n",
      "terminase\n",
      "lipoyl synthase\n",
      "glucosaminyl deacetylase\n",
      "rnasee\n",
      "gtp cyclohydrolase i\n",
      "\n",
      "\n",
      "GROUP  4\n",
      "5' nucleotidase\n",
      "phosphatase\n",
      "oxidoreductase\n",
      "deoxynucleoside monophosphate kinase\n",
      "dna binding, hu-like domain\n",
      "\n",
      "\n",
      "GROUP  5\n",
      "pnuc-like nicotinamide riboside transporter\n",
      "histidine triad nucleotide binding protein\n",
      "nicotinate ribosyltransferase\n",
      "galactosyltransferase\n",
      "dna ligase\n",
      "\n",
      "\n",
      "GROUP  6\n",
      "scaffolding protein\n",
      "metalloprotease\n",
      "protease\n",
      "muf-like minor capsid protein\n",
      "capsid decoration protein, lamd-like\n",
      "\n",
      "\n",
      "GROUP  7\n",
      "head-to-tail adaptor\n",
      "head-to-tail stopper\n",
      "minor capsid protein\n",
      "tail terminator\n",
      "major capsid and protease fusion protein\n",
      "\n",
      "\n",
      "GROUP  8\n",
      "minor tail protein, d-ala-d-ala carboxypeptidase\n",
      "nucleoside deoxyribosyltransferase\n",
      "peptidyl trna hydrolase\n",
      "laglidadg endonuclease\n",
      "\n",
      "\n",
      "GROUP  9\n",
      "dnae-like dna polymerase iii (alpha)\n",
      "dna polymerase iii sliding clamp (beta)\n",
      "fabg-like reductase\n",
      "rf-1 peptide chain release factor\n",
      "\n",
      "\n",
      "GROUP  10\n",
      "tail sheath protein\n",
      "baseplate j protein\n",
      "baseplate wedge protein\n",
      "tail tube protein\n",
      "\n",
      "\n",
      "GROUP  11\n",
      "tellurium resistance protein d family\n",
      "dnac-like helicase loader\n",
      "repa-like replication initiator\n",
      "reca-like dna recombinase\n",
      "\n",
      "\n",
      "GROUP  12\n",
      "rna binding protein\n",
      "aminotransferase\n",
      "mutt-like nucleotide pyrophosphohydrolase\n",
      "thymidylate kinase\n",
      "\n",
      "\n",
      "GROUP  13\n",
      "adp-ribosyltransferase\n",
      "capsid maturation protease and muf-like fusion protein\n",
      "udp-glucose dehydrogenase\n",
      "rna ligase\n",
      "\n",
      "\n",
      "GROUP  14\n",
      "adenylyltransferase\n",
      "3' end\n",
      "acetyltransferase\n",
      "thioesterase\n",
      "\n",
      "\n",
      "GROUP  15\n",
      "dna polymerase i\n",
      "kinase\n",
      "adenylosuccinate synthetase, pura-like\n",
      "\n",
      "\n",
      "GROUP  16\n",
      "aaa-atpase\n",
      "ku-like dsdna break-binding protein\n",
      "cobt-like cobalamin biosynthesis protein\n",
      "\n",
      "\n",
      "GROUP  17\n",
      "excise\n",
      "nucleoside deaminase\n",
      "pyrimidine deaminase\n",
      "\n",
      "\n",
      "GROUP  18\n",
      "cas4 family exonuclease\n",
      "imm-like superinfection immunity protein\n",
      "recb-like exonuclease/helicase\n",
      "\n",
      "\n",
      "GROUP  19\n",
      "nuclease\n",
      "dna primase/helicase\n",
      "repa-like helicase\n",
      "\n",
      "\n",
      "GROUP  20\n",
      "rexa family abortive infection protein\n",
      "rexb family abortive infection protein\n",
      "peptidase\n",
      "\n",
      "\n",
      "GROUP  21\n",
      "recombination directionality factor\n",
      "mre11 double-strand break endo/exonuclease\n",
      "rna polymerase sigma factor\n",
      "\n",
      "\n",
      "GROUP  22\n",
      "rece-like exonuclease\n",
      "rect-like ss dna binding protein\n",
      "rusa-like resolvase (endonuclease)\n",
      "\n",
      "\n",
      "GROUP  23\n",
      "serine/threonine kinase\n",
      "glycosyltransferase\n",
      "n-acetylglucosaminyltransferase\n",
      "\n",
      "\n",
      "GROUP  24\n",
      "serine integrase\n",
      "histidine kinase\n",
      "sprt-like protease\n",
      "\n",
      "\n",
      "GROUP  25\n",
      "dna primase\n",
      "chaperonin, dnaj-like\n",
      "ssdna binding protein, erf family\n",
      "\n",
      "\n",
      "GROUP  26\n",
      "tail needle protein\n",
      "lysin a\n",
      "tailspike protein\n",
      "\n",
      "\n",
      "GROUP  27\n",
      "lipoprotein\n",
      "lysm-like peptidoglycan binding protein\n",
      "tape measure protein\n",
      "\n",
      "\n",
      "GROUP  28\n",
      "thymidylate synthase\n",
      "dihydrofolate reductase\n",
      "phosphoribosyl transferase\n",
      "\n",
      "\n",
      "GROUP  29\n",
      "ftsk-like dna translocase\n",
      "nucleotidyl transferase\n",
      "trna nucleotidyltransferase\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "groupings = greedy_modularity_communities(G_markov, weight=None)\n",
    "\n",
    "for g in range(30):\n",
    "    print(\"GROUP \",g)\n",
    "    for i in groupings[g]:\n",
    "        print(labels[i])\n",
    "    print()\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
